pygsti/tools/optools.py

```diff
@@ -450,20 +450,51 @@ def entanglement_fidelity(a, b, mx_basis='pp'):
         The basis of the matrices.  Allowed values are Matrix-unit (std),
         Gell-Mann (gm), Pauli-product (pp), and Qutrit (qt)
         (or a custom basis object).
+        
+    is_tp_flag : bool, optional (default None)
+        Manual flag for specifying that both matrices are TP. Skips
+        the check for this which is faster, but only should be used
+        when user is certain this is true apriori.
+    is_unitary_flag : bool, optional (default None)
+        Manual flag for specifying that the second matrix, b, is
+        unitary. This skips the check for this which is faster, but only
+        should be used when the user is certain this is true apriori.
 
     Returns
     -------
     float
     """
     d2 = a.shape[0]
-    def is_tp(x): return _np.isclose(x[0, 0], 1.0) and all(
-        [_np.isclose(x[0, i], 0) for i in range(d2)])
-
-    def is_unitary(x): return _np.allclose(_np.identity(d2, 'd'), _np.dot(x, x.conjugate().T))
-
-    if is_tp(a) and is_tp(b) and is_unitary(b):  # then assume TP-like gates & use simpler formula
-        TrLambda = _np.trace(_np.dot(a, b.conjugate().T))  # same as using _np.linalg.inv(b)
-        d2 = a.shape[0]
+    
+    
+
+    
+
+    #if the tp flag isn't set we'll calculate whether it is true here
+    if is_tp_flag is None:
+        def is_tp(x): return _np.isclose(x[0, 0], 1.0) and all(
+        [_np.isclose(x[0, i], 0) for i in range(1,d2)])
+        
+        is_tp_flag= (is_tp(a) and is_tp(b))
+   
+    #if the unitary flag isn't set we'll calculate whether it is true here 
+    if is_unitary_flag is None:
+        def is_unitary(x): return _np.allclose(_np.identity(d2, 'd'), _np.dot(x, x.conjugate().T))
+        
+        is_unitary_flag= is_unitary(b)
+    
+        
+    if is_tp_flag and is_unitary_flag:  # then assume TP-like gates & use simpler formula
+        #old version, slower than einsum
+        #TrLambda = _np.trace(_np.dot(a, b.conjugate().T))  # same as using _np.linalg.inv(b)
+        
+        #Use einsum black magic to only calculate the diagonal elements
+        #if the basis is either pp or qm we know the elements are real-valued, so we
+        #don't need to take the conjugate
+        if mx_basis=='pp' or mx_basis=='qm':
```

Comment:
Is this line correct?  From the comment, which jives with my understanding, .conjugate() can be *dropped* when the basis is 'pp' or 'qm', not *added*.  I think this may be a "sign" error ;)

Reply:
Also, while we're using einsum magic, we could swap `ji` to `ij` and drop the `.T` in the three `einsum` lines, e.g.  `_np.einsum('ij,ij->',a, b)`.  This may be even slightly faster?

Reply:
Thanks for catching that, they were indeed switched. Incidentally, I also had a typo in the basis label for the gell-mann matrices I hadn't noticed.

As for dropping the transpose, I just gave that a shot and found that it gives a 7% speedup for the 2-qubit case, but is within error bars of the version with the transpose for the 3-qubit case. Since numpy implements transposes using views, I'm guessing that the overhead of constructing that view is still measurable for 2-qubits, but is negligible for 3. Given that I'm going to opt to keep the transpose in for readability.